### PR TITLE
Performance improvements for parsers

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

**Comment by harendra-kumar** [comment]
Benchmark improvements:

```
Data.Parser(cpuTime)
Benchmark                                                                       default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------- -------------- --------------------------

All.Data.Parser/o-1-space.takeFramedByEsc_                                             1642.10                      -4.14
All.Data.Parser/o-n-heap.someAlt                                                       3857.19                      -5.31
All.Data.Parser/o-1-space.sepByAll (words)                                              261.23                      -7.31
All.Data.Parser/o-1-space.concatSequence                                               4529.62                     -15.98
All.Data.Parser/o-1-space.deintercalate1                                               1138.14                     -18.55
All.Data.Parser/o-1-space.deintercalate                                                1169.63                     -20.87
All.Data.Parser/o-1-space.sepBy1 (words)                                               1059.96                     -25.73
All.Data.Parser/o-1-space.sepBy (words)                                                1104.38                     -28.56
All.Data.Parser/o-1-space.manyTill                                                     3714.10                    -323.49
All.Data.Parser/o-1-space.splitApAfter                                                  471.36                    -514.04
All.Data.Parser/o-1-space.splitAp                                                       489.05                    -534.24
All.Data.Parser/o-1-space.splitWith                                                     495.47                    -543.46
All.Data.Parser/o-n-heap.split_                                                         448.37                    -683.10
All.Data.Parser/o-1-space.splitApBefore                                                 460.60                    -713.79
All.Data.Parser/o-n-heap.lookAhead                                                     1414.74                   -3667.88
All.Data.Parser/o-1-space.takeEQ                                                       1462.87                   -3764.26
All.Data.Parser/o-1-space.streamEqBy                                                   7405.70                   -8382.12
```

File: core/src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 1]
```diff
@@ -2231,6 +2226,15 @@ streamEqBy cmp (D.Stream sstep state) = Parser step initial extract
 
     extract _ = return $ Error "streamEqBy: end of input"
 
+-- | Like 'listEqBy' but uses a stream instead of a list and does not return
+-- the stream.
+--
+{-# INLINE streamEqBy #-}
+streamEqBy :: Monad m => (a -> a -> Bool) -> D.Stream m a -> Parser a m ()
+-- XXX Somehow composing this with "*>" is much faster on the microbenchmark.
+-- Need to investigate why.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserD.hs:2235]: How did you even figure this out?

File: benchmark/Streamly/Benchmark/Data/Parser.hs [hunk 2]
```diff
@@ -567,3 +573,9 @@ main = do
             (o_1_space_serial_unfold value arrays)
         , bgroup (o_n_heap_prefix moduleName) (o_n_heap_serial value)
         ]
+#else
+    let value = 100000
+    let input = sourceUnfoldrM value 1
+    manyTill value input
+    return ()
+#endif
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Parser.hs:592]: This looks like some test code. Wildly specific.

File: core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs [hunk 3]
```diff
@@ -434,37 +369,92 @@ parseDToK
...
-                    (cont (Success n b) (count + n - off) (Chunk arr))
-            ChunkPartial n pst1 ->
-                assert (n < 0 || n >= Array.length arr)
-                    (return $ Partial n (parseCont (count + n - off) pst1))
-            ChunkContinue n pst1 ->
-                assert (n < 0 || n >= Array.length arr)
-                    (return $ Continue n (parseCont (count + n - off) pst1))
-            ChunkError n err ->
-                cont (Failure n err) (count + n - off) (Chunk arr)
+    parseContChunk !count !offset !state arr@(Array contents start end) = do
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserK/Type.hs:387]: Did not review this in depth.

**Review by adithyaov — APPROVED** [review]

**Pull request closed without merging**